pinot-query-planner/src/main/java/org/apache/pinot/query/routing/table/TableRouteProvider.java

```diff
@@ -0,0 +1,135 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.routing.table;
+
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.request.BrokerRequest;
+import org.apache.pinot.core.routing.RoutingManager;
+import org.apache.pinot.core.routing.TimeBoundaryInfo;
+import org.apache.pinot.core.transport.TableRouteInfo;
+import org.apache.pinot.spi.config.table.TableConfig;
+
+
+/**
+ * The TableRoute interface provides the metadata required to route query execution to servers. The important sources
+ * of the metadata are table config, broker routing information and the broker request.
+ */
+public interface TableRouteProvider {
```

Comment:
Currently the functionality of this interface is mixed with `TableRouteInfo`, where both of them contains routing info.
Ideally this interface should encapsulate the logic of computing routing info, and should be re-usable across queries. All the routing info should be stored within `TableRouteInfo`. With the current PR, half of the info are stored in the provider and half stored in `TableRouteInfo`.
Can you please try to split the computing logic and result, and only keep the logic in the provider (maybe rename it to computer), and the info in `TableRouteInfo`? One principal to help you with the split is to see if you can reuse the computing logic across queries.

Reply:
Done. Please check now

Reply:
Much better now. Do we still need 2 methods? Shall we merge it into 1?

Reply:
It has to be in two separate functions because the control flow goes back and forth between the route provider and `BaseSingleStageBrokerRequestHandler`. This conversation is highly relevant: https://github.com/apache/pinot/pull/15388#discussion_r2056924754

Reply:
Also smaller functions make better unit testing possible. Each of these functions have 100+ unit tests each